### PR TITLE
Issue 1059 add _in_memory to AlignTraj

### LINKED ISSUE #1059: AlignTraj should have in_memory option
### Expected behaviour
With the introduction of MemoryReader in 0.16, `rms_fit_trj` gained an in_memory option, which makes it possible to do in-place (in-memory) alignment. `AlignTraj` should have the same option.

### Currently version of MDAnalysis:
0.16.0-dev0

### PR DESCRIPTION
**Pull request by orbeckst**
Fixes #1059

Changes made in this Pull Request:
 - add NullWriter
 - in_memory=True flag for AlignTraj

PR Checklist
------------
 - [x] does it work?
 - [x] Tests?
 - [x] Docs?
 - **na** because feature was introduced this cycle CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

File: package/MDAnalysis/analysis/align.py [hunk 1]
```diff
@@ -512,12 +512,26 @@ def __init__(self, mobile, reference, select='all', filename=None,
             Last frame of trajectory to analyse, Default: -1
         step : int, optional
             Step between frames to analyse, Default: 1
+        in_memory : boolean, optional
+            *Permanently* switch `mobile` to an in-memory trajectory
+            so that alignment can be done in-place, which can improve
+            performance substantially in some cases. In this case, no file
+            is written out (`filename` and `prefix` are ignored) and only
+            the coordinates of `mobile` are changed in memory.
+        Default:
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/align.py]: You don't need to specify the default here. The original function signature with all default values will be shown in docs.

File: package/MDAnalysis/analysis/align.py [hunk 2]
```diff
@@ -528,23 +542,32 @@ def __init__(self, mobile, reference, select='all', filename=None,
...
 
-        if filename is None:
-            path, fn = os.path.split(self._trajectory.filename)
-            filename = os.path.join(path, prefix + fn)
-            logger.info('filename of rms_align with no filename given'
-                        ': {0}'.format(filename))
+        if in_memory or isinstance(mobile.trajectory, MemoryReader):
+            mobile.transfer_to_memory()
+            filename = None
+            logger.info("Moved mobile trajectory to in-memory representation")
```
Comment (kain88-de) [inline package/MDAnalysis/analysis/align.py]: No need to log if we already have a memory trajectory

File: package/MDAnalysis/coordinates/null.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+# MDAnalysis: A Toolkit for the Analysis of Molecular Dynamics Simulations.
+# J. Comput. Chem. 32 (2011), 2319--2327, doi:10.1002/jcc.21787
+#
+"""
+Null output --- :mod:`MDAnalysis.coordinates.null`
+==================================================
+
+The :class:`NullWriter` provides a Writer instance that behaves like
+any other writer but effectively ignores its input and does not write
+any output files. This is similar to writing to ``/dev/null``.
```
Comment (kain88-de) [inline package/MDAnalysis/coordinates/null.py]: You should add that this class exists to help writing generic analysis code that works with in-memory and normal trajectories.

**Comment by orbeckst** [comment]
@kain88-de thanks for the feedback, I will fix docs, but the main problem is that it doesn't work yet.

**Comment by orbeckst** [comment]
@wouterboomsma – comments are welcome on this PR which is attempting to add `in_memory` to `AlignTraj`. At the moment, it does not seem that the in-memory mobile universe is properly superimposed. After running 

```python
import MDAnalysis as mda
from MDAnalysis.tests.datafiles import PSF, DCD
from MDAnalysis.analysis.align import AlignTraj

ref = mda.Universe(PSF, DCD)
ref.trajectory[-1]
u = mda.Universe(PSF, DCD)

x = AlignTraj(u, ref, in_memory=True).run()
print(x.rmsd)
```
the `rmsd` appears to be the same for all frames.
```
[ 6.92913669  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669
  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669
  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669
  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669
  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669
  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669
  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669
  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669
  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669
  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669
  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669
  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669
  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669
  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669
  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669
  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669  6.92913669
  6.92913669  6.92913669]
```

When doing `in_memory=False` I get something sensible:
```python
ref = mda.Universe(PSF, DCD)
ref.trajectory[-1]
u = mda.Universe(PSF, DCD)
y = AlignTraj(u, ref, in_memory=False).run()
print(y.rmsd)
```

```
[  6.92913669e+00   6.83231394e+00   6.72506086e+00   6.66130556e+00
   6.59454998e+00   6.50351078e+00   6.41261003e+00   6.35668639e+00
   6.29181505e+00   6.20077437e+00   6.11645360e+00   6.05047761e+00
   5.94797693e+00   5.89244206e+00   5.78170470e+00   5.69452321e+00
   5.59754764e+00   5.51804911e+00   5.43079593e+00   5.37585526e+00
   5.29992552e+00   5.22743772e+00   5.15543970e+00   5.10918650e+00
   5.03625227e+00   4.95363924e+00   4.89721900e+00   4.81169501e+00
   4.73502328e+00   4.66882647e+00   4.59721735e+00   4.55930164e+00
   4.47875296e+00   4.41940427e+00   4.37317802e+00   4.28618127e+00
   4.19828079e+00   4.12897269e+00   4.04636917e+00   3.96897425e+00
   3.88579962e+00   3.80973668e+00   3.73946990e+00   3.67540856e+00
   3.59324884e+00   3.52478213e+00   3.45778394e+00   3.38134804e+00
   3.28929794e+00   3.23041626e+00   3.16338284e+00   3.09563980e+00
   3.02501307e+00   2.95364919e+00   2.88826405e+00   2.80037035e+00
   2.73312894e+00   2.66643042e+00   2.61508745e+00   2.54567025e+00
   2.48339337e+00   2.42153505e+00   2.33747553e+00   2.27371495e+00
   2.19279801e+00   2.12547451e+00   2.08860889e+00   2.03351084e+00
   1.96231782e+00   1.90915904e+00   1.86117727e+00   1.82757700e+00
   1.78012388e+00   1.74149884e+00   1.72176216e+00   1.67015665e+00
   1.64703238e+00   1.59329660e+00   1.56346583e+00   1.53780447e+00
   1.50856975e+00   1.44711872e+00   1.38387962e+00   1.31278403e+00
   1.26792523e+00   1.18382992e+00   1.15025136e+00   1.09318645e+00
   1.03325532e+00   9.69927406e-01   9.20878786e-01   8.63770272e-01
   8.08022081e-01   7.41313761e-01   6.58074602e-01   6.10992569e-01
   5.46916762e-01   7.46666772e-07]
```

(I am not sure when I will have time again to hack away here so contributions from anyone are certainly welcome, too.)

**Comment by kain88-de** [comment]
@orbeckst I think I fixed it and another problem. The current implementation would only write the selection we used for alignment and not the whole topology.  I'll check further though.

@wouterboomsma it looks like the memory reader has a problem when I only update the part of the underlying array using a selection. I'll look a bit more into that later

**Comment by kain88-de** [comment]
OK now I found the error. The problem was that we assigned the old Reader class to the base for iterating before we did the conversion to a memory reader. Now we first convert the reader to be in_memory and then call the super class init function.

**Comment by kain88-de** [comment]
> it looks like the memory reader has a problem when I only update the part of the underlying array using a selection. I'll look a bit more into that later

This is also invalid. It works just fine.

File: package/MDAnalysis/analysis/align.py [hunk 4]
```diff
@@ -534,16 +534,12 @@ def __init__(self, mobile, reference, select='all', filename=None,
...
-        if self._quiet:
-            logging.disable(logging.WARN)
-
         select = rms.process_selection(select)
         self.ref_atoms = reference.select_atoms(*select['reference'])
         self.mobile_atoms = mobile.select_atoms(*select['mobile'])
-
         if in_memory or isinstance(mobile.trajectory, MemoryReader):
             mobile.transfer_to_memory()
+            print("in memory")
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/align.py]: This commit is currently for debugging? Otherwise remove the `print`.

**Comment by kain88-de** [comment]
I fixed all issues with develop now and my own comments. There is also a test now for the partial fit bug I found while looking at this. From my side this is done.

File: package/MDAnalysis/coordinates/null.py (hunk 3 above)
Comment (orbeckst) [inline package/MDAnalysis/coordinates/null.py]: Thanks for adding the docs.

**Comment by orbeckst** [comment]
@kain88-de thanks for the AlignTraj fixes. 

I am now looking at the Travis failures – probably related to the NullWriter. I am working on my local branch where I rebased against develop and merged some of the commits.

**Comment by kain88-de** [comment]
I've restarted the build conda seems to have timed out on travis

**Comment by wouterboomsma** [comment]
@orbeckst, sorry for the lack of response. I've been away. Does anything more need to be done here? I guess the Encore code should be updated to use `AlignTraj` instead of `rms_fit_trj`, which would be another test that everything works as it should. Shall I do this in a separate PR?

**Comment by orbeckst** [comment]
@wouterboomsma can you please update the Encore code in a separate PR and also  the blog post? Thanks!

--
Oliver Beckstein
email: orbeckst@gmail.com

**Comment by wouterboomsma** [comment]
@orbeckst, @mtiberti has agreed to make the necessary changes in encore over the next days.

**Pull request merged**